source/validate.cpp

```diff
@@ -215,13 +226,52 @@ spv_result_t ValidateBinaryUsingContextAndValidationState(
 
   // NOTE: Parse the module and perform inline validation checks. These
   // checks do not require the the knowledge of the whole module.
-  if (auto error = spvBinaryParse(&context, vstate, words, num_words,
-                                  setHeader, ProcessInstruction, pDiagnostic))
+  if (auto error = spvBinaryParse(&context, vstate, words, num_words, setHeader,
+                                  ProcessInstruction, pDiagnostic)) {
     return error;
+  }
+
+  // NOTE: Copy each instruction for easier processing
+  std::vector<spv_instruction_t> instructions;
+  uint64_t index = SPV_INDEX_INSTRUCTION;
+  while (index < binary->wordCount) {
+    uint16_t wordCount;
+    uint16_t opcode;
+    spvOpcodeSplit(spvFixWord(binary->code[index], endian), &wordCount,
+                   &opcode);
+    spv_instruction_t inst;
+    spvInstructionCopy(&binary->code[index], static_cast<SpvOp>(opcode),
+                       wordCount, endian, &inst);
+    instructions.push_back(inst);
+    index += wordCount;
+  }
 
-  if (vstate->in_function_body())
+  // If unrecognized extensions are used and 'permissive' mode is enabled,
+  // consider the module valid.
+  if (context.permissive) {
```

Comment:
You need more than this anyway.  The full spvBinaryParse is done above at line 229, and that will do what it calls "inline checks".  If you really want to exit early, then you'll have to put the check for unknown extension in the middle of that parsing and set a bit in the validation state.  Also, the ProcessInstruction function should exit early with success if it sees that bit set.